Azure Devops & Github Actions error logging

ArmaanMcleod opened the pull request:
<!-- Thank you for contributing to Pester! -->

## PR Summary
<!-- Please describe what your pull request fixes, or how it improves Pester.

If your pull request resolves a reported issue, please mention it by using `Fix #<issue_number>` on a new line, this will close the linked issue automatically when this PR is merged. For more info see: [Closing issues using keywords](https://help.github.com/articles/closing-issues-using-keywords/).

If your pull request integrates Pester with another system, please tell us how the change can be tested. -->

Fix #1364.

I've added functionality to highlight CI errors in red for ADO and GithubActions. 

This is configured from the `Output.CIFormat`, which covers the following options:

- **None** -> No CI autodetection
- **Auto** -> Autodetects CI
- **AzureDevops**
- **GithubActions**

For AzureDevops, we check the `TF_BUILD` environment variable is equal to `True`.
For GithubActions, we check the `GITHUB_ACTIONS` environment variable is equal to `True`. 

For **AzureDevops** we can use `##vso[task.logissue type=error]` to highlight the error header, and `##[error]` to highlight the messages. This ensures only the error header is reported in the build log. 

For **GithubActions**, you can only set an error, so I had to limit this to just the header, as shown below. The messages are also collapsed under the header using expandable groups.

### Azure Devops Output

![ADO](https://user-images.githubusercontent.com/20082136/122637604-fc83d400-d132-11eb-88a0-ede87b9ec5c8.PNG)

### Build Log

![ADOBuildLog](https://user-images.githubusercontent.com/20082136/122637786-df9bd080-d133-11eb-93c2-fb0a0f20e62b.PNG)


### Github Actions

![GithubActions](https://user-images.githubusercontent.com/20082136/122943004-fe12fd80-d3b9-11eb-93a3-0206bec674b6.PNG)

### Build Log

![GithubActionsBuildLog](https://user-images.githubusercontent.com/20082136/122637969-b29bed80-d134-11eb-8096-88e8d895edc7.PNG)

I did testing in my own repo by running a simple github action to test out this code: https://github.com/ArmaanMcleod/TestGithubActions/actions/runs/952195814

## PR Checklist

- [x] PR has meaningful title
- [x] Summary describes changes
- [x] PR is ready to be merged
  - If not, use the arrow next to `Create Pull Request` to mark it as a draft. PR can be marked `Ready for review` when it's ready.
- [x] Tests are added/update *(if required)*
- [x] Documentation is updated/added *(if required)*

<!-- Before you continue, please review [Contributing to Pester](https://pester.dev/docs/contributing/introduction).

Our continuous integration system doesn't send any notifications about failed tests. Please return to the opened pull request (after ~60 minutes) to check if everything is OK. -->

ArmaanMcleod commented:
I guess another problem is that this floods the Azure Devops logs with error messages now. 
![buildlog](https://user-images.githubusercontent.com/20082136/122250299-e3014300-cf0c-11eb-96e5-3af489661179.PNG)

ArmaanMcleod commented:
I will need to add some tests for this as well.

nohwnd commented on src/Main.ps1:
```diff
@@ -926,6 +926,10 @@ function Invoke-Pester {
                 $PesterPreference.Output.StackTraceVerbosity = "Full"
             }
 
+            if ([Environment]::GetEnvironmentVariable("TF_BUILD") -eq 'True') {
```
```suggestion
            # Variable is set to 'True' if the script is being run by a build task. https://docs.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml
            # Do not fix this to check for boolean value, the value is set to literal string 'True'
            if ([Environment]::GetEnvironmentVariable("TF_BUILD") -eq 'True') {
```

nohwnd commented on src/Pester.RSpec.ps1:
```diff
@@ -398,6 +398,9 @@ function New-PesterConfiguration {
       StackTraceVerbosity: The verbosity of stacktrace output, options are None, FirstLine, Filtered and Full.
       Default value: 'Filtered'
+
+      FormatAzureDevopsErrorMessages: Formats Azure Devops error messages with red color.
```
```suggestion
      FormatAzureDevopsErrorMessages: Formats Azure Devops error messages with red color. This is automatically set to true when AzureDevOps run is detected, by checking that TF_BUILD environment variable is set to 'True'. https://docs.microsoft.com/en-us/azure/devops/pipelines/build/variables
```

nohwnd reviewed (commented):
It might be nicer if the option was more generic, azdo can do also warning and info (IIRC). Even if this PR would implement just the errors, having just a single option for enabling the azdo formatted output would be better.

We should also ensure that user can override this in config to be disabled even if the env variable is set. Or vice versa.

nohwnd commented:
Changed this to be draft, because it is work in progress.

ArmaanMcleod commented:
> It might be nicer if the option was more generic, azdo can do also warning and info (IIRC). Even if this PR would implement just the errors, having just a single option for enabling the azdo formatted output would be better.
> 
> We should also ensure that user can override this in config to be disabled even if the env variable is set. Or vice versa.

Thanks for the feedback. Agree fully about supporting other log types. 

For overriding the option to `$false` even if `TF_BUILD` exists, should we include another option to ignore this environment variable? Thinking that might be easier, since the option is set to `$false` by default, and it would be difficult to determine how the env variable would get ignored without something else to override it. Perhaps I'm overcomplicating it, let me know your thoughts 👍

nohwnd commented:
There is `IsOriginalValue()` on the options on config for this purpose. 

```powershell
PS> $c = New-PesterConfiguration
PS> $c.CodeCoverage.Enabled.Value -eq $false
True
PS> $c.CodeCoverage.Enabled.IsOriginalValue()
True
PS> $c.CodeCoverage.Enabled = $false
PS> $c.CodeCoverage.Enabled.IsOriginalValue()
False
```

I look at this during the merging, so this is probably automatic, I think the setting to pester preference happens before merging it to user provided config, and in that case user provided value will take precedence if they set it explicitly.

nohwnd commented:
`FormatAzureDevopsLogMessages`

Sounds complicated, give me some time to come up with some better name. @fflaten do you have any opinion?

ArmaanMcleod commented:
Ah that's perfect, `IsOrignalValue()` will definitely solve the issue 👍 .

Yeah I don't like `FormatAzureDevopsLogMessages` either now that I'm using it. Happy to suggestions for a better name. Perhaps one improvement would be to replace `Format` with `Write`. Could also shorten `AzureDevops` to something like `ADO` or `AzDO`.

fflaten commented:
Do we need a ADO-specific option? Today it's Azure, tomorrow it's GitHub Actions (if we can use https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions from Pester), Travis CI etc. I'm not familiar all the different CI/CD systems. 

Would a enum like Output.CIFormat be better? Since it modifies the output while supporting only one value should be supported at a time (None being default).

fflaten commented:
Just tested, GitHub Actions format works:
![image](https://user-images.githubusercontent.com/3436158/122444541-f231d080-cfa0-11eb-839b-fcce9a3d2216.png)

So I'd personally expect that to be the next step as ADO (Pester's own CI) and GitHub Actions are equally popular.

fflaten commented on src/functions/Output.ps1:
```diff
@@ -773,6 +829,25 @@ function Get-WriteScreenPlugin ($Verbosity) {
+        $LogType
+    )
+
+    foreach ($line in $Message) {
+
+        # https://docs.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=powershell#logissue-log-an-error-or-warning
+        & $SafeCommands['Write-Host'] "##vso[task.logissue type=$LogType]$line"
+    }
```
I think we should skip the foreach and only append to first line, typically `[-] mypath.to.failing.test`. The error flags are used in the report, so appending the special prefix to every message will fill the report with noise for the same error. Ex.
![image](https://user-images.githubusercontent.com/3436158/122450100-e9dc9400-cfa6-11eb-947c-77d28e8e9565.png)

fflaten commented:
Can we limit the prefix to the first line only, typically the failing block/test (`[-] mypath.to.failing.test`)?
The error flags are used in the pipeline report as shortcuts to the log, so appending the special prefix to every message will fill the report with noise for the same error. Ex.
![image](https://user-images.githubusercontent.com/3436158/122450100-e9dc9400-cfa6-11eb-947c-77d28e8e9565.png)

ArmaanMcleod commented:
> Would a enum like Output.CIFormat be better? Since it modifies the output while supporting only one value should be supported at a time (None being default).

I'm happy to create a `Output.CIFormat` option instead, which would be more generic for multiple CI providers. I'm thinking we can just support ADO and Github Actions like you said. 

I'm also wondering what levels of output would we support?

ADO supports the following:

```pwsh
Write-Host "##[section] This is colored green!"
Write-Host "##[command] This is colored blue!"
Write-Host "##[debug] This is colored purple!"
Write-Host "##[warning] This is colored orange!"
Write-Host "##vso[task.logissue type=warning;]This is colored orange!" 
Write-Host "##[error] This is colored red!"
Write-Host "##vso[task.logissue type=error;]This is colored red!"
```

Which looks like this:

<img width="120" alt="CIOutput" src="https://user-images.githubusercontent.com/20082136/122495364-3c609380-d02d-11eb-9a17-a6769d2bcb8b.PNG">

For now it seems errors and warnings should definitely be supported. Should we support the other format types as well? Like Debug for instance. 

> Can we limit the prefix to the first line only, typically the failing block/test (`[-] mypath.to.failing.test`)?
> The error flags are used in the pipeline report as shortcuts to the log, so appending the special prefix to every message will fill the report with noise for the same error. Ex.

Also agree on this. The error logs get flooded with what I've currently done and limiting just the first line makes it less noisy.

nohwnd commented:
Option name `CIColor` ?


Coloring with just a single issue: `Write-Host "##vso[task.logissue type=error;]This is colored red!"` This for first line and then the color for the rest? `Write-Host "##[error] This is colored red!"`

Is original value, will probably by set after merging on the $PesterPreference, that we do few lines above. You need to move the code before  `& $SafeCommands['Get-Variable'] 'Configuration' -Scope Local | Remove-Variable` and check the value that user provided. And also the $callerPreference from the context. If either of those are set we should not overwrite with automatic value.

nohwnd commented:
Levels: I would support only error for now.

fflaten commented:
> Option name `CIColor` ?

Isn't this limiting? Color itself is better handled with ANSI as it's generic and supported in many CI-systems. 

The added benefit with this type of support is the integration with CI-summary/reports.

nohwnd commented:
AutodetectCI and specify what that means for each CI in the help?

ArmaanMcleod commented:
> AutodetectCI and specify what that means for each CI in the help?

So this autodetect option would be set to `true` by default? I guess that means disabling it might be simpler now since user can override to `false` and it won't autodetect. Of course even if its enabled we would still need to check if the env variable exists, and if it doesn't exist we set it to `false` and output a warning that no CI is detected? This seems like a simple solution to me.

fflaten commented:
> Is original value, will probably by set after merging on the $PesterPreference, that we do few lines above. You need to move the code before `& $SafeCommands['Get-Variable'] 'Configuration' -Scope Local | Remove-Variable` and check the value that user provided. And also the $callerPreference from the context. If either of those are set we should not overwrite with automatic value.

If the merged configuration falsely reports IsOriginalValue=false for options even though it wasn't modified in neither Configuration nor callerPreference, then that sounds more like a bug in Merger imho. I haven't tested it, but I'd expect it to not lie even after merging considering it only intends to modified already user-modified settings. We include it in the result-object so users should be able to detect modified/default options if they want.

> AutodetectCI and specify what that means for each CI in the help?

Sounds like it might conflict with the `-CI` switch some people are used to.  I'd like it to be a bit more explicit, but I won't argue anymore on this.

Personally I like 
```
Output.CIFormat = 'Auto' (default), 'None', 'AzureDevOps', 'GithubActions'
# As StringOption or EnumOption (to get faster validation of false input)

# Auto-detect if not specified
if ($PesterPreference.Output.CIFormat.IsOriginalValue()) {
    if ($env:thatAzDOdetectionThing) {
        $PesterPreference.Output.CIFormat = 'AzureDevOps'
    }
    elseif ($env:MagicGitHubActionsDetectionCondition {
        $PesterPreference.Output.CIFormat = 'GithubActions'
    }
}
```

fflaten commented:
TODO right before merging:
Disable in PesterPreference for RSpec tests since they fail on purpose.

ArmaanMcleod commented:
> TODO right before merging:
> Disable in PesterPreference for RSpec tests that fail on purpose.

Thanks @fflaten, I added some commits to disable this for deliberate failing tests. If I have to test and/or add more more changes I will stash these commits and apply them later. 

@nohwnd This is probably ready for review again. I've updated the PR description with the new changes.

Let me know what you guys think 🙂 .

ArmaanMcleod commented on src/Main.ps1:
```diff
@@ -899,6 +899,32 @@ function Invoke-Pester {
                 [PesterConfiguration] $PesterPreference = [PesterConfiguration]::Merge($callerPreference, $Configuration)
             }
 
+            if ($PesterPreference.Output.CIFormat.Value -eq 'Auto') {
```
I didn't check `IsOriginalValue` here since its `Auto` by default. This will work for when its manually set or left by default.

ArmaanMcleod commented on src/Pester.RSpec.ps1:
```diff
@@ -399,6 +399,9 @@ function New-PesterConfiguration {
       StackTraceVerbosity: The verbosity of stacktrace output, options are None, FirstLine, Filtered and Full.
       Default value: 'Filtered'
 
+      CIFormat: The CI format of output in build logs, options are None, Auto, AzureDevops and GithubActions. Environment variable TF_BUILD(https://docs.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml#system-variables-devops-services) is automically set to true when Azure Devops is detected. Environment variable GITHUB_ACTIONS(https://docs.github.com/en/acti ...
```
I've added documentation here explaining what those environment vars do and why they are set to `True`. Let me know if I should be more explicit.

fflaten replied:
I'd keep this short and simple + accepted values. How we auto-detect and how it looks should be added to https://pester.dev. For example in https://github.com/pester/docs/pull/128

ArmaanMcleod commented on src/functions/Output.ps1:
```diff
@@ -773,6 +829,69 @@ function Get-WriteScreenPlugin ($Verbosity) {
+            # message error prefx, but doesn't get reported in build log. Only task issue errors get reported.
+            MessagePrefix = '##[error] '
+        }
+        'GithubActions' = @{
```
Seems a bit hacky to format like this for GithubActions, but seems to align header with rest of message ok for now.

fflaten commented on src/Main.ps1:
```diff
@@ -899,6 +899,32 @@ function Invoke-Pester {
+                }
+            }
+
+            elseif ($PesterPreference.Output.CIFormat.Value -eq 'AzureDevops' -and $env:TF_BUILD -ne 'True') {
+                $PesterPreference.Output.CIFormat = 'None'
+            }
+
+            elseif ($PesterPreference.Output.CIFormat.Value -eq 'GithubActions' -and $env:GITHUB_ACTIONS -ne 'True') {
+                $PesterPreference.Output.CIFormat = 'None'
+            }
+
```
The intention is good, but I would remove this. If the environment variables changes in the future, it's okay if auto-detect breaks, but explicit user settings should not break.

What do you think?

ArmaanMcleod replied:
Yeah I see your point. I think we probably just remove `AzureDevops` and `GithubActions` options and keep just `Auto` and `None`. In that case might it better to just use a boolean flag like `AutoDetectCI` as we discussed before?

ArmaanMcleod replied:
On the other hand keeping explicit CI options like this helps with determining what CI formatting to apply. I'd assume if we did `AutoDetectCI`, I would still need another property to set the CI that has been detected, since I currently pass this in into a hashtable with the CI header/message formatting.

ArmaanMcleod replied:
I've removed these lines and the associated tests.

fflaten reviewed (commented):
Looks good 👍 Added some thoughts

ArmaanMcleod commented:
I've also realised I need to add this to `Main.ps1`, since there is a call to `Write-ErrorToScreen` in there that shows exceptions to the screen, which need to be shown in CI as well. Not sure why I missed that. 

Will mark this as draft and work on that 🙂

nohwnd commented:
I will have another look on this later this week.

ArmaanMcleod commented:
I made some changes to the GHA formatting, where the messages under the error header are collapsed into an expandable group. 

Screenshots are in the PR description.

nohwnd commented on tst/Pester.RSpec.Configuration.ts.ps1:
```diff
@@ -623,11 +627,247 @@ i -PassThru:$PassThru {
+            $r = Invoke-Pester -Configuration $c
+            $r.Configuration.Output.CIFormat.Value | Verify-Equal "AzureDevops"
+
+            $env:TF_BUILD = $previousTfBuildVariable
```
Those would not be set back if the test fails. You should use try finally to wrap the call to Invoke-Pester.

ArmaanMcleod replied:
Ah good catch, thanks for spotting that 👍 . I've wrapped those calls in `try/finally`.

nohwnd commented:
The output looks great, and I am reasonably happy about the naming (can't come up with better one though). Only a small nit pick about the variables in tests, and then it's imho ready to merge. Will wait for the changes, and @fflaten approve.

fflaten commented:
Great work @ArmaanMcleod ! 🎉 

I've been busy the last days, but I didn't have anything to add in a review. It's mostly the option-design/name I'm not sure about, but I think we've settled on this for now. Maybe we get some feedback from other users with the now released alpha. 👍
